File: app/src/main/java/com/example/tradeup/data/repository/UserSavedItemsRepositoryImpl.java
package com.example.tradeup.data.repository;

import androidx.annotation.NonNull;
import com.example.tradeup.core.utils.Callback;
import com.example.tradeup.data.source.remote.FirebaseUserSavedItemsSource;

import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class UserSavedItemsRepositoryImpl implements UserSavedItemsRepository {

    private final FirebaseUserSavedItemsSource firebaseUserSavedItemsSource;

    @Inject
    public UserSavedItemsRepositoryImpl(FirebaseUserSavedItemsSource firebaseUserSavedItemsSource) {
        this.firebaseUserSavedItemsSource = firebaseUserSavedItemsSource;
    }

    @Override
    public void saveItem(@NonNull String userId, @NonNull String itemId, final Callback<Void> callback) {
        firebaseUserSavedItemsSource.saveItem(userId, itemId)
                .addOnSuccessListener(aVoid -> callback.onSuccess(null))
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void unsaveItem(@NonNull String userId, @NonNull String itemId, final Callback<Void> callback) {
        firebaseUserSavedItemsSource.unsaveItem(userId, itemId)
                .addOnSuccessListener(aVoid -> callback.onSuccess(null))
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getSavedItemIds(@NonNull String userId, final Callback<List<String>> callback) {
        firebaseUserSavedItemsSource.getSavedItemIds(userId)
                .addOnSuccessListener(callback::onSuccess)
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void isItemSaved(@NonNull String userId, @NonNull String itemId, final Callback<Boolean> callback) {
        firebaseUserSavedItemsSource.isItemSaved(userId, itemId)
                .addOnSuccessListener(callback::onSuccess)
                .addOnFailureListener(callback::onFailure);
    }
}

File: app/src/main/java/com/example/tradeup/data/repository/UserSavedItemsRepositoryImpl.kt
package com.example.tradeup.data.repository

import com.example.tradeup.data.source.remote.FirebaseUserSavedItemsSource // Giả sử bạn sẽ tạo FirebaseUserSavedItemsSource.kt
// import com.example.tradeup.data.model.Item // Nếu cần
import javax.inject.Inject

class UserSavedItemsRepositoryImpl @Inject constructor(
    private val firebaseUserSavedItemsSource: FirebaseUserSavedItemsSource
) : UserSavedItemsRepository {

    override suspend fun saveItem(userId: String, itemId: String): Result<Unit> {
        // return firebaseUserSavedItemsSource.saveItem(userId, itemId)
        TODO("Implement saveItem in FirebaseUserSavedItemsSource and call it here")
    }

    override suspend fun unsaveItem(userId: String, itemId: String): Result<Unit> {
        // return firebaseUserSavedItemsSource.unsaveItem(userId, itemId)
        TODO("Implement unsaveItem in FirebaseUserSavedItemsSource and call it here")
    }

    override suspend fun getSavedItemIds(userId: String): Result<List<String>> {
        // return firebaseUserSavedItemsSource.getSavedItemIds(userId)
        TODO("Implement getSavedItemIds in FirebaseUserSavedItemsSource and call it here")
    }

    override suspend fun isItemSaved(userId: String, itemId: String): Result<Boolean> {
        // return firebaseUserSavedItemsSource.isItemSaved(userId, itemId)
        TODO("Implement isItemSaved in FirebaseUserSavedItemsSource and call it here")
    }
    // ...
}